### PR TITLE
[Statistics] Behavioural Statistics > Double Data Entry - Site permission fix

### PR DESCRIPTION
**Pull request by laemtl**
### Description of the issue
In Statistics / Behavioural Statistics / Double Data Entry Statistics, users were allowed to see the double data entry breakdown per participant from sites they don't have access to. 
### Brief summary of changes
The `_hasAccess` function was incorpoerate to this class (same functionality that the one in #5591)

The function `_checkCriteria` was modified for only retrieve data from the sites(centers) the user have access to.

Note: In future developments the capacity of setting permissions per site per user could be desirable. A note in rapport was included in the codeset. In this case the proposed function `_hasAccess` should be updated accordingly .

#### Testing instructions (if applicable)
A user with permission 'data_entry' should be now able to access the 'breakdown per participant' for the double data entry page only for the sites it have access to.

#### Links to related PRs
#5950
#5591
Duplicates of #5966 (rebased on 23.0)

### CODE REVIEW

**Comment by laemtl** [comment]
I re tested this rebase and it works as expected.

File: modules/statistics/php/statistics_dd_site.class.inc [hunk 1]
```diff
@@ -39,16 +89,53 @@ class Statistics_DD_Site extends statistics_site
...
+        // The filter _checkCriteria() takes care of restricting
+        // the user access only to the sites it belongs to.
+        // When logic reimplemented on hasCenterPermission(),
+        // _checkCriteria() will take care of retriving information
+        // only for those centers the user has the specific permission.
         if (!empty($centerID)) {
             $this->query_criteria   .= " AND s.CenterID =:cid ";
             $this->query_vars['cid'] = $centerID;
+        } else {
+            $list_of_permitted_sites = (array) null;
```
Comment (maltheism) [inline modules/statistics/php/statistics_dd_site.class.inc:109]: Interesting way to make an empty array. I've never seen this way until today. 🙂

**Comment by maltheism** [comment]
I'm going to manually test. The code looks good!

**Comment by maltheism** [comment]
Hi @laemtl here are my results from what my test user could see visible. The General Demographics Statistics seems to be correct with only showing the two projects the user has access. The other statistics tabs in the module show all the projects for the select field.







User test permissions:

**Comment by laemtl** [comment]
Thank you @maltheism. I will apply the same fix to the other sections as well!

**Comment by maltheism** [comment]
I think this PR will fix #6675 @kongtiaowang

**Comment by kongtiaowang** [comment]
@maltheism This PR will Behavioural Statistics. But Imaging Statistics has the same issue.

**Comment by christinerogers** [comment]
@laemtl What's the status?   Is this ready for @maltheism's re-review?  If so, please also ask Karim and/or Alex to help you test and review this today.

**Comment by laemtl** [comment]
@christinerogers Al the comments here were addressed and merge with #6706

**Comment by laemtl** [comment]
@AlexandraLivadas @h-karim Does any of you have time to test this as well?

**Comment by christinerogers** [comment]
@driusan - final review pls

**Comment by laemtl** [comment]
@christinerogers very good suggestion. I created #6742 and #6743.

**Comment by ridz1208** [comment]
@pierre-p-s please review and test

**Comment by pierre-p-s** [comment]
The behavioural statistics tab seems to be working for all sites except for 'Data Coordinating Center' (DCC). When you have the permission to access only DCC data, nothing appears in the behavioural statistics tab (see screenshot below). I am not getting any error in the console or error log. Is this intended?

Screen Shot 2020-07-03 at 12 23 31 PM

When clicking on breakdown per participant with only DCC permission this is what I am getting:

Screen Shot 2020-07-03 at 12 26 40 PM

**Comment by driusan** [comment]
DCC is not included in statistics by design.

**Comment by ridz1208** [comment]
@pierre-p-s DCC is not included in stats by design

File: modules/statistics/php/statistics_dd_site.class.inc [hunk 2]
```diff
@@ -46,14 +46,9 @@ class Statistics_DD_Site extends statistics_site
...
 
-        // TODO: There are no means of set permissions per site
-        // for a given user right now: (e.g.) The user X can have
-        // the permission data_entry on site Y but not on site Z.
-        // Currently, hasCenterPermission() function is only checking
-        // if the user have a given center AND a given permission
-        // not if it have the permission for this specific center.
-        // This logic will be implemented in hasCenterPermission()
-        // in near versions when the permission framework allow it.
+        // TODO (#6742): There are no means of set permissions per site
```
Comment (christinerogers) [inline modules/statistics/php/statistics_dd_site.class.inc:49]:
```suggestion
        // TODO (#6742): There are no means of setting permissions per site
```

File: modules/statistics/php/statistics_dd_site.class.inc [hunk 3]
```diff
@@ -62,14 +57,8 @@ class Statistics_DD_Site extends statistics_site
                 intval($_REQUEST['CenterID'])
             );
         } else {
-            // For the short term the user we'll be granted access
+            // FIXME (#6742): For the short term the user we'll be granted access
```
Comment (christinerogers) [inline modules/statistics/php/statistics_dd_site.class.inc:60]:
```suggestion
            // FIXME (#6742): For the short term the user will be granted access
```

File: modules/statistics/php/statistics_dd_site.class.inc [hunk 4]
```diff
@@ -62,14 +57,8 @@ class Statistics_DD_Site extends statistics_site
...
             // if at least have permission AND one of the centers
```
Comment (christinerogers) [inline modules/statistics/php/statistics_dd_site.class.inc:61]:
```suggestion
            // if they have permission and belong to ANY of the centers
```
Edit to clarify what the issue is -- the problematic part is that any center affiliation get them all centers.

File: modules/statistics/php/statistics_dd_site.class.inc [hunk 5]
```diff
@@ -89,19 +79,14 @@ class Statistics_DD_Site extends statistics_site
...
-        // for a given user right now: (e.g.) The user X can have
-        // the permission data_entry on site Y but not on site Z.
-        // Currently, hasCenterPermission() function is only checking
-        // if the user have a given center AND a given permission
-        // not if it have the permission for this specific center.
-        // This logic will be implemented in hasCenterPermission()
-        // in near versions when the permission framework allow it
-        // The filter _checkCriteria() takes care of restricting
-        // the user access only to the sites it belongs to.
+        // TODO (#6743): _checkCriteria takes care of restricting access
```
Comment (christinerogers) [inline modules/statistics/php/statistics_dd_site.class.inc:82]:
```suggestion
        // TODO (#6743): _checkCriteria takes care of restricting data access
```
clarifying that it's data access not e.g. page-loading access

File: modules/statistics/php/statistics_dd_site.class.inc [hunk 6]
```diff
@@ -89,19 +79,14 @@ class Statistics_DD_Site extends statistics_site
...
+        // to sites the user belongs to.
         // When logic reimplemented on hasCenterPermission(),
-        // _checkCriteria() will take care of retriving information
+        // _checkCriteria() will take care of retreiving information
```
Comment (christinerogers) [inline modules/statistics/php/statistics_dd_site.class.inc:85]:
```suggestion
        // _checkCriteria() will take care of retrieving information
```

File: modules/statistics/php/statistics_dd_site.class.inc [hunk 7]
```diff
@@ -89,19 +79,14 @@ class Statistics_DD_Site extends statistics_site
...
         // only for those centers the user has the specific permission.
+
+        // TODO (#6743): There are no means of set permissions per site
```
Comment (christinerogers) [inline modules/statistics/php/statistics_dd_site.class.inc:88]:
```suggestion
        // TODO (#6743): There are no means of setting permissions per site
```

File: modules/statistics/php/statistics_dd_site.class.inc [hunk 8]
```diff
@@ -25,9 +25,49 @@ namespace LORIS\statistics;
...
+    /**
+     * Checking user's permission
+     *
+     * @param \User $user The user whose access is being checked
+     *
+     * @return bool
+     */
+    function _hasAccess(\User $user) : bool
+    {
+        // TODO (#6742): Create a permission specific to statistics
```
Comment (pierre-p-s) [inline modules/statistics/php/statistics_dd_site.class.inc:39]: I feel like this should not be in the code but rather should be documented in the PR. Same thing for all the other mentions of issues.

**Comment by laemtl** [comment]
Restarted the build and it succeeded.

File: modules/statistics/php/statistics_dd_site.class.inc [hunk 9]
```diff
@@ -39,9 +79,42 @@ class Statistics_DD_Site extends statistics_site
...
             $this->query_criteria   .= " AND s.CenterID =:cid ";
             $this->query_vars['cid'] = $centerID;
+        } else {
+            $list_of_permitted_sites = (array) null;
+            $currentUser = \User::singleton();
+            if ($currentUser->hasPermission('access_all_profiles')) {
+                $list_of_permitted_sites = array_keys(\Utility::getSiteList());
+            } else {
+                foreach ($currentUser->getCenterIDs() as $centerID) {
+                    if ($currentUser->hasCenterPermission(
```
[the remark this answers is not in the document]
Comment (laemtl) [reply modules/statistics/php/statistics_dd_site.class.inc:100]: All the suggestions make sense. I can't comment on the code as I didn't write it. Maybe @racostas can comment.

File: modules/statistics/php/statistics_dd_site.class.inc [hunk 10]
```diff
@@ -25,9 +25,47 @@ namespace LORIS\statistics;
...
+        // data_entry permission at the site/center specified by CenterID.
+        if (!empty($_REQUEST['CenterID'])) {
+            $hasCenterPermission = $user->hasCenterPermission(
+                'data_entry',
+                intval($_REQUEST['CenterID'])
+            );
+        } else {
+            // FIXME (#6742): For the short term the user will be granted access
+            // if they have permission for a minimum of one of the centers
+            foreach ($user->getCenterIDs() as $centerID) {
```
Comment (driusan) [inline modules/statistics/php/statistics_dd_site.class.inc:60]: Comment from 6 days ago.. not sure why GitHub removed it:

I don't understand this check.. under what situations would the user have the centerID returned from getCenterIDs but not have permission to the centerID?

Comment (ridz1208) [reply modules/statistics/php/statistics_dd_site.class.inc:60]: @racostas can you comment on @driusan 's question ?

Comment (laemtl) [reply modules/statistics/php/statistics_dd_site.class.inc:60]: Additional information can be found here: https://github.com/aces/Loris/issues/6742.
Part of the comment has been migrated in an issue.

Comment (racostas) [reply modules/statistics/php/statistics_dd_site.class.inc:60]: Hi, thanks for bring this back. This code is a bit old now I think. I will take a look but sure there are lot of things could be refactored/fixed.

**Comment by racostas** [comment]
#6861 refactors the code in this PR. Also includes permission restrictions by project . Please refer to this new version.

**Comment by racostas** [comment]
closed by #6861

**Pull request closed without merging**